[Yeap Rayson] iP

Respirayson opened the pull request:
# Respironix

> It’s not a bug; it’s an undocumented feature. ― Anonymous

Respironix frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

## What do you need to do????

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :shipit:
And it is **FREE!**

## Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Respirayson committed:
Add rename, greet and exit

Respirayson committed:
add echo functionality

Respirayson committed:
add add, list functionality

Respirayson committed:
add mark as done functionality

Respirayson committed:
add todos, events and deadlines functionality

Respirayson committed:
added the input.txt and expected.txt

Respirayson committed:
add new exceptions and add logic

Respirayson committed:
add delete functionality

Respirayson committed:
converted commands to enums

Respirayson committed:
Add logic to read file from storage

Respirayson committed:
Add exceptions and transfer content from file

Respirayson committed:
Change backslash to File.separator

Respirayson committed:
Add write file functionality

Respirayson committed:
Delete unnecessary imports

Respirayson committed:
Merge branch 'branch-Level-7'

Respirayson committed:
Create formatter for localdatetime

Respirayson committed:
Add logic to store and parse localdatetime

Respirayson committed:
Update text-ui-testing

Respirayson committed:
Add exception to handle backwards timeframe

Respirayson committed:
Add command to print tasks on specific date

Respirayson committed:
Add more format types for datetime

Respirayson committed:
Merge branch 'branch-Level-8'

Respirayson committed:
Add TaskList

Respirayson committed:
Add Storage

Respirayson committed:
Add ui

Respirayson committed:
Add showAdd, showOutro and move printTasks into ui

Respirayson committed:
Add more functions into ui

Respirayson committed:
Add showing error messages

Respirayson committed:
Add Commands and Parser

Respirayson committed:
Modify Duke

Respirayson committed:
Fix no line found exception due to scanner

Respirayson committed:
Merge branch 'branch-A-MoreOOP'

Respirayson committed:
Refactor code into packages

Respirayson committed:
Fix text-ui-test bug

Respirayson committed:
Merge branch 'A-Packages'

Respirayson committed:
Configure checkstyle

Respirayson committed:
Merge branch 'add-gradle-support'

Respirayson committed:
Add DeadlineTest and modify Deadline

Respirayson committed:
Add EventTest and modify DeadlineTest

Respirayson committed:
Merge branch 'branch-A-JUnit'

Respirayson committed:
Add JavaDoc for Commands

Respirayson committed:
Add JavaDoc for tools

Respirayson committed:
Rename packages

Respirayson committed:
Remove unnecessary this

Respirayson committed:
Remove this in Duke class

Respirayson committed:
Fix import statements

Respirayson committed:
Fix more checkstyle violations

Respirayson committed:
Add Javadoc for exceptions

Respirayson committed:
Add Find Command and logic

Respirayson committed:
Add javadoc for find

Respirayson committed:
Merge branch 'branch-A-CodingStandard'

Respirayson committed:
Merge branch 'branch-A-JavaDoc'

Respirayson committed:
Fix bug for gradle run

Respirayson committed:
Merge branch branch-Level-9

Respirayson committed:
Fix import bugs

Respirayson committed:
Fix checkstyle violations

Respirayson committed:
Fix checkstyle and text-ui-testing bugs

Respirayson committed:
Change build and write first program

Respirayson committed:
Add GUI using fxml

anthonytamzil commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,72 @@
+public class Duke {
+
+    // CONSTANTS
+    private static final String DIR_NAME = "data";
+    private static final String FILE_NAME = "duke.txt";
```
I like how you used all uppercases for your constant names

anthonytamzil commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,111 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws NoDescriptionException,
+            UnknownTimeException, BackwardsTimeException, UnknownCommandException {
+        Parser.Operations operation = Parser.Operations.valueOf(
+                fullCommand.replaceAll("^\\W*\\b(\\w+).*", "$1").toUpperCase());
```
I like that you adhered to the indentation rules for wrapped lines!

anthonytamzil commented on src/main/java/duke/tools/Storage.java:
```diff
@@ -0,0 +1,119 @@
+package duke.tools;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+import duke.exceptions.UnrecognisedFormatException;
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.ToDo;
```
I like the consistent ordering of your import statements!

anthonytamzil commented on src/main/java/duke/tasks/ToDo.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public String formatForStorage() {
+        return String.format("T | %s", super.formatForStorage());
+    }
```
Perhaps you can add some documentation here too?

anthonytamzil reviewed (commented):
Looks absolutely fantastic.
Just a few nits from me, that's all.

awhb commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,111 @@
+package duke.commands;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+import duke.exceptions.BackwardsTimeException;
+import duke.exceptions.NoDescriptionException;
+import duke.exceptions.UnknownCommandException;
+import duke.exceptions.UnknownTimeException;
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.ToDo;
+import duke.tools.Parser;
+import duke.tools.Storage;
+import duke.tools.TaskList;
+import duke.tools.Ui;
```
I like how the duke imports are organised alphabetically! As an extension, you could also consider ordering the java imports after the duke imports!

awhb commented on src/test/java/duke/tasks/EventTest.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Test
+    public void testStringConversion() {
```
A note that according to the Java coding standard, names like "stringConversion" would suffice as test methods for the method it is meant to test e.g. "stringConversion"

awhb commented on src/main/java/duke/tools/Parser.java:
```diff
@@ -0,0 +1,74 @@
+        // Use Regex to extract the first word even with preceding whitespace
+        Operations command = Operations.valueOf(fullCommand.replaceAll("^\\W*\\b(\\w+).*", "$1").toUpperCase());
+
+        switch (command) {
+        case TODO:
+            // Fallthrough
+        case DEADLINE:
+            // Fallthrough
+        case EVENT:
+            return new AddCommand(fullCommand, FORMATTER);
+        case BYE:
+            return new ExitCommand();
+        case LIST:
+            return new ListCommand();
+        case MARK:
+            return new MarkCommand(fullCommand);
+        case UNMARK:
+            return new UnmarkCommand(fullCommand);
+        case DELETE:
+            return new DeleteCommand(fullCommand);
+        case CHECK:
+            return new CheckCommand(fullCommand, FORMATTER);
+        case FIND:
+            return new FindCommand(fullCommand);
+        default:
+            throw new UnknownCommandException(fullCommand);
+        }
```
Great work avoiding a common coding violation of indenting the switch cases!

awhb commented on src/main/java/duke/tools/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                    throw new UnrecognisedFormatException();
+                }
+
+                boolean isDone = segments[1].equals("1");
```
Great work naming your booleans identifiably according to the Java coding standard consistently throughout your code!

awhb reviewed (commented):
Great work 👍 
Minor nits recommended

Respirayson committed:
Modify Ui for GUI

Respirayson committed:
Add Javadoc and fix checkstyle

Respirayson committed:
Change Javadoc name

Respirayson committed:
Merge branch 'branch-Level-10'

Respirayson committed:
Amend Javadoc

Respirayson committed:
Change colour of button

Respirayson committed:
Merge branch 'branch-GUI-improvements'

Respirayson committed:
Modify styles for Javafx

Respirayson committed:
Create main.yml

Respirayson committed:
Add assertions to tools package

Introduce assertions to enhance the robustness and reliability of the
code in the classes in the tools package.

Assertions need to be strategically placed in the code to ensure
various conditions and invariants are met during program execution.

Let's ensure
- Add file existence assertion to confirm the presence
  of the storage file before loading tasks.
- Add format assertions to check for the format of the storage file.
- Apply length assertions to the tasks to ensure their format.
- Add file write assertion to confirm the success of the write operation
  to the storage file.

These assertions will assist in identifying potential issues during
development and debugging phases, ensuring the code behaves as
expected and meets the specified requirements.

Respirayson committed:
Add remaining Assertions

Need to ensure that certain methods are not called with null arguments.

Calling methods with null arguments will result in a
NullPointerException.

Lets,
* add assertions to ensure that the arguments passed
  to the methods are not null.
* add assertions to check for empty strings.

Respirayson committed:
Fix checkstyle violations

Respirayson committed:
Merge pull request #1 from Respirayson/branch-A-Assertions

Add assertions into the code

Respirayson committed:
Add comments for readability

There are several methods where I performed regex matching. Without
comments, it is hard to understand what the regex is doing.

Adding comments to explain the regex helps to improve code quality
and readability.

Lets,
* Add comments to explain the regex in Parser.java
* Add comments to explain the regex in Storage.java
* Add comments to explain the regex in Ui.java
* Add comments to explain the regex in all the Commands.java

Respirayson committed:
Fix merge conflicts

Respirayson committed:
Merge pull request #2 from Respirayson/branch-A-CodeQuality

Improve Code Quality

Respirayson committed:
Add basic archiving functionality

Respirayson committed:
Ensure data is persisted after each command

Respirayson committed:
Merge pull request #3 from Respirayson/branch-C-Archive

BCD-Extension

Respirayson committed:
Abstract task creation from AddCommand to Task

Respirayson committed:
Refactor error catching in Storage

Respirayson committed:
Refactor Storage

Respirayson committed:
Fix checkstyle violations

Respirayson committed:
Merge branch 'branch-codeQuality-feedback'

Respirayson committed:
Add rounded profile icons

Respirayson committed:
Change font

Respirayson committed:
Add padding and change spacing and colour of GUI

Respirayson committed:
Refactor code to pass checkstyle

Respirayson committed:
Add personality to error handling

Respirayson committed:
Merge branch 'branch-A-BetterGui'

Respirayson committed:
Add starting phrase to List

Respirayson committed:
Merge branch 'branch-A-Personality'

Respirayson committed:
Add Ui.png

Respirayson committed:
Improve error handling for enums

Respirayson committed:
Add more conversational touches to UI

Respirayson committed:
Change name of add and delete message methods

Respirayson committed:
Update docs/Readme.md

Respirayson committed:
Merge branch 'branch-A-UserGuide'

Respirayson committed:
Move operator to newline for checkstyle

Respirayson committed:
Edit _config.yml to render gfmd table

Respirayson committed:
Update emoji to unicode to support jar file
